org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/PreferenceManager.java

```diff
@@ -68,6 +70,7 @@
 	private ListenerList<IPreferencesChangeListener> preferencesChangeListeners;
 	private IEclipsePreferences eclipsePrefs;
 	private static Map<String, Template> templates = new LinkedHashMap<>();
+	private static Collection<IPath> triggerFiles;
```

Comment:
We should not persist the trigger files. They should be used only once. And only use them to calculate a default source path when initializing a new invisible project.

If the user modifies the source paths in a subsequent action, such as changing the workspace setting `java.project.sourcePaths`, we should always follow the user's action to update the source paths.

This means that if the user sets sourcePaths to null, we should just clean up all the source paths. There is no need to fall back and use trigger files to calculate the source paths again. Since the user knows the settings to change the source paths, he/she can change them back if he/she wants.